File: JAVA_study_note/java_one_object.java
//一个类只生成一个对象,java表示
//注：java中无参数的的函数不能用yy（void）形式
class A
{
	public int i;

	private static A aa = new A();//在类中生成一个对象。aa只是引用（C/C++中指针），一个指向该类类型的一快内存，对象在堆中分配
								//设置成私有为了保护成员，防止不小心破坏结构
	public static A getA()//该函数必须设成静态的，和公共的，保证在类外能访问，而且不能属于对象。
	{
		return aa;//返回在类中生成的引用地址，这样在类外也能访问在类中生成的那个对象
	}
	private A(){}//把构造函数设置成私有的，外部就不可以访问，也就没发直接定义该类的对象
}

class java_one_object
{
	public static void main (String[] args)
	{
		//A aa = new A();//错误，类的构造函数是私有的，类外不可访问
		A aa = A.getA();//把在类中生成的对象的地址赋给aa
		A bb = A.getA();//把在类中生成的对象的地址赋给bb,此时aa和bb相等
		
		if(aa == bb)
		{
			System.out.printf ("aa = bb\n");
		}
		else 
		{
			System.out.printf ("aa != bb\n");
		}
		
		aa.i = 100;
		System.out.printf ("aa.i=%d  bb.i=%d\n",aa.i,bb.i);
	}
}

File: JAVA_study_note/java_one_object.cpp
//一个类只生成一个对象，C++表示
/*C++中完全用指针实现，不推荐使用，弊病太大···、

C++中的堆中的管理完全是可以控制的，这就导致了另一个问题，如果多个指针同时指向这块内存，如果用其中一个指针释放了这块内存，
其他指针在访问时，就会导致程序崩溃····
*/
#include <iostream>

using namespace std;

class A
{
	A(){}//把构造函数设置成私有的，外部就不可以访问，也就没发直接定义该类的对象
	static A * aa;//声明一个指向该类的指针
public:
	int i;
	static A * getA()
	{
		return aa;
	}
};

A * A::aa = new A;//定义和初始化aa，在堆中分配一个对象空间

int main ()
{
	//A * aa = new A();//错误，类的构造函数是私有的，类外不可访问
	//A aa;//错误，类的构造函数是私有的，类外不可访问
		A * aa = A::getA();//把在类中生成的对象的地址赋给aa
		A * bb = A::getA();//把在类中生成的对象的地址赋给bb,此时aa和bb相等
		
		if(aa == bb)
		{
			cout<< "aa = bb\n";
		}
		else 
		{
			cout<< "aa != bb\n";
		}
		
		aa->i = 100;
		cout <<"aa.i="<< aa->i<< "  "<<"bb.i="<<bb->i<<endl;
	
}